api/unstable/fingerprint.h

```diff
@@ -35,6 +35,102 @@ typedef enum {
     S2N_FINGERPRINT_JA3,
 } s2n_fingerprint_type;
 
+struct s2n_fingerprint;
+
+/**
+ * Create a reusable fingerprint structure.
+ *
+ * Fingerprinting is primarily used to identify malicious or abusive clients,
+ * so fingerprinting needs to be efficient and require minimal resources.
+ * The `s2n_client_hello_get_fingerprint_hash` and `s2n_client_hello_get_fingerprint_string`
+ * methods may require additional memory to calculate the fingerprint. Reusing
+ * the same `s2n_fingerprint` structure to calculate multiple fingerprints reduces
+ * the cost of each individual fingerprint.
+ *
+ * @param type The algorithm to use for the fingerprint.
+ * @returns S2N_SUCCESS on success, S2N_FAILURE on failure.
+ */
+S2N_API struct s2n_fingerprint *s2n_fingerprint_new(s2n_fingerprint_type type);
+
+/**
+ * Frees the memory allocated by `s2n_fingerprint_new` for a fingerprint structure.
+ *
+ * @param fingerprint The fingerprint structure to be freed.
+ * @returns S2N_SUCCESS on success, S2N_FAILURE on failure.
+ */
+S2N_API int s2n_fingerprint_free(struct s2n_fingerprint **fingerprint);
+
+/**
+ * Resets the fingerprint for safe reuse with a different ClientHello.
+ *
+ * @param fingerprint The fingerprint structure to be reset.
+ * @returns S2N_SUCCESS on success, S2N_FAILURE on failure.
+ */
+S2N_API int s2n_fingerprint_reset(struct s2n_fingerprint *fingerprint);
+
+/**
+ * Sets the ClientHello to be fingerprinted.
+ *
+ * @param fingerprint The fingerprint to be modified
+ * @param ch The client hello to be fingerprinted. It will not be copied, so needs
+ * to live at least as long as this fingerprinting operation.
+ * @returns S2N_SUCCESS on success, S2N_FAILURE on failure.
+ */
+S2N_API int s2n_fingerprint_set_client_hello(struct s2n_fingerprint *fingerprint, struct s2n_client_hello *ch);
+
+/**
+ * Get the size of the fingerprint hash.
```

Comment:
```suggestion
 * Get the size of the hex-encoded fingerprint hash.
```
I think it's useful to clarify, otherwise I have a tiny question in the back of my mind about whether I need to multiply this by 2 to get enough space for the hex-encoded `s2n_fingerprint_get_hash`

Reply:
That actually brings up a good point that I could use some opinions on:

I don't want to write "hex-encoded" because JA4 is only partially hex encoded. Here's a JA4 example: "t13i310900_e8f1e7e78f70_1f22a2ca17c4". Notice it starts with a prefix that isn't hex, then has two different truncated hashes that are hex.

But along the same vein, is "hash" even an accurate name for it anymore?! I considered just "fingerprint", but then the C looks like "s2n_fingerprint_get_fingerprint" which is kind of silly. I could leave the C as "s2n_fingerprint_get"? I also considered "short", so "s2n_fingerprint_get_short", but that doesn't seem great either. It doesn't have a name in Wireshark or the documentation-- it's just "JA3" or "JA4".

The longer version was called "full" / "fullstring" in JA3 but is now "raw" / "r" in JA4. I went with "raw" because I like it better without strong reasoning, but maybe there are other opinions on that too.

We basically have two values: one which is a known, reasonable length, and one which is variable length and can be unreasonably long. Based on how fingerprints are used, I think that separation will remain constant across methods, even though the exact format and contents of the strings will keep changing. Naming the two fields is just hard.

Reply:
I think I'd prefer `short` over `hash`? It's less descriptive but I'd expect that to be to our benefit as we add more fingerprint types, e.g. if the short representation of some fingerprint in the future doesn't actually rely on a hash.

Reply:
I'm actually changing my mind here. While trying to think of more, better names for a fixed-size version of a variable length string, Google told me "hash". From [wikipedia](https://en.wikipedia.org/wiki/Hash_function): "A hash function is any function that can be used to map data of arbitrary size to fixed-size values, though there are some hash functions that support variable length output." I guess we forget that more general definition because we always deal with cryptographic hash functions. But by that definition, even the JA4 string is a "hash", and "hash" is exactly the right name for this.

Reply:
I'm still a tiny bit worried that other people also mostly work with cryptographic hashes and would get confused, but if that feedback surfaces publicly it should be pretty easily solved with additional documentation or examples. Since we have the JA-3 hash form included in the doc comments, it should be pretty straightforward for any questioning users to unblock themselves.